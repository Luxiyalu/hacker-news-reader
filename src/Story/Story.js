import React from 'react';
import base from '../config/base';
import './Story.scss';

class Story extends React.PureComponent {
  constructor(props) {
    super(props);
    this.state = {
      story: {}
    };
  }

  componentDidMount() {
    base.fetch(`/v0/item/${this.props.id}`, {
      context: this,
      then(story) {
        this.setState({ story });
      }
    });
  }

  timeSince(timestamp) {
    const date = new Date(timestamp * 1000);
    const seconds = Math.floor((new Date() - date) / 1000);
    let interval = Math.floor(seconds / 31536000);

    if (interval > 1) return interval + ' years ago';
    interval = Math.floor(seconds / 2592000);
    if (interval > 1) return interval + ' months ago';
    interval = Math.floor(seconds / 86400);
    if (interval > 1) return interval + ' days ago';
    interval = Math.floor(seconds / 3600);
    if (interval > 1) return interval + ' hours ago';
    interval = Math.floor(seconds / 60);
    if (interval > 1) return interval + ' minutes ago';
    return 'just now';
  }

  render() {
    const { story } = this.state;

    // No need to go through render if data is not fetched
    if (!Object.keys(story).length) return null;

    return (
      <div className="Story">
        <a href={story.url}>{story.title}</a>
        <p>
          <span>
            {story.score} {story.score <= 1 ? 'point' : 'points'}
          </span>
          <span> by {story.by} </span>
          <span>{this.timeSince(story.time)}</span>
        </p>
      </div>
    );
  }
}

export default Story;
