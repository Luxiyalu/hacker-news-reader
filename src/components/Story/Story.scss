.story-container {
  padding: 14px 20px;
  transition-duration: 1s;
  transition-timing-function: cubic-bezier(0, 0, 0.2, 1);
  transition-property: background;

  &.fresh {
    background: rgba(255, 102, 0, 0.2);
  }

  a {
    text-decoration: none;
    color: black;
    font-weight: 500;
  }

  p {
    margin-top: 5px;
    color: grey;
    font-size: 14px;
  }
}
