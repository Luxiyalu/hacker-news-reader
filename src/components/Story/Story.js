import React from 'react';
import { base, storage, timeSince } from '../../services/';
import './Story.scss';

class Story extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      story: {}
    };
  }

  componentDidMount() {
    const { id } = this.props;
    const storedStory = storage.getStoryById(id);

    if (storedStory && Object.keys(storedStory).length) {
      this.setState({ story: storedStory });
    } else {
      base.fetch(`/v0/item/${id}`, {
        context: this,
        then(story) {
          // if fetch result is empty, endpoint has not updated yet. wait 500ms and try again.
          if (story && Object.keys(story).length) {
            this.setState({ story });
            storage.setStory(id, story);
          } else {
            setTimeout(this.componentDidMount.bind(this), 500);
          }
        }
      });
    }
  }

  render() {
    const { story } = this.state;

    // No need to go through render if data is not fetched
    if (!Object.keys(story).length) return null;

    return (
      <div className="Story">
        <a href={story.url}>{story.title}</a>
        <p>
          <span>
            {story.score} {story.score <= 1 ? 'point' : 'points'}
          </span>
          <span> by {story.by} </span>
          <span>{story.time} </span>
          <span>{timeSince(story.time)}</span>
        </p>
      </div>
    );
  }
}

export default Story;
