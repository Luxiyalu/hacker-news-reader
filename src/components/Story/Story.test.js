import React from 'react';
import Story from './Story';
import renderer from 'react-test-renderer';

it('renders without crashing', () => {
  const component = renderer.create(<Story key={1} id={1} />);
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

it('initiate with the correct init state', () => {});

it('fetch story data after component mount', () => {});

it('caches at localstorage to prevent duplicated fetches', () => {});
