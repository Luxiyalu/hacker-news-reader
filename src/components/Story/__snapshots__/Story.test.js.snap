// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Story /> renders without crashing 1`] = `null`;
